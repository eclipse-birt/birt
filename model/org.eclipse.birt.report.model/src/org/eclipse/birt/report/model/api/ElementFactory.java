/*******************************************************************************
 * Copyright (c) 2004 Actuate Corporation.
 * 
 * This program and the accompanying materials are made available under the
 * terms of the Eclipse Public License 2.0 which is available at
 * https://www.eclipse.org/legal/epl-2.0/.
 * 
 * SPDX-License-Identifier: EPL-2.0
 * 
 *
 * Contributors:
 *  Actuate Corporation  - initial API and implementation
 *******************************************************************************/

package org.eclipse.birt.report.model.api;

import org.eclipse.birt.report.model.core.Module;

/**
 * Creates a new report elements and returns handles to it. Use this to create
 * elements. After creating an element, add it to the design using the
 * <code>add</code> method in the {@link SlotHandle}class. Obtain an instance of
 * this class by calling the <code>getElementFactory</code> method on any
 * element handle.
 * 
 * @see SlotHandle
 */

public class ElementFactory extends ElementFactoryImpl {

	/**
	 * Constructs a element factory with the given module.
	 * 
	 * @param module the module
	 */

	public ElementFactory(Module module) {
		super(module);
	}

}
