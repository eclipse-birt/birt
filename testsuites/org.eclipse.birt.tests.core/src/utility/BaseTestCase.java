package utility;
import java.io.File;
import java.net.URL;
import java.security.CodeSource;
import java.security.ProtectionDomain;
import java.util.ArrayList;
import java.util.List;

import junit.framework.TestCase;

/*******************************************************************************
 * Copyright (c) 2004 Actuate Corporation. All rights reserved. This program and
 * the accompanying materials are made available under the terms of the Eclipse
 * Public License v1.0 which accompanies this distribution, and is available at
 * http://www.eclipse.org/legal/epl-v10.html
 * 
 * Contributors: Actuate Corporation - initial API and implementation
 ******************************************************************************/

public class BaseTestCase extends TestCase
{

	protected static final String TEST_FOLDER = "src/"; //$NON-NLS-1$

	protected String getTestFolder( )
	{
		return TEST_FOLDER;
	}

	/**
	 * Locates the folder where the unit test java source file is saved.
	 * 
	 * @return the path name where the test java source file locates.
	 */

	protected String getClassFolder( )
	{

		String pathBase = null;

		ProtectionDomain domain = this.getClass( ).getProtectionDomain( );
		if ( domain != null )
		{
			CodeSource source = domain.getCodeSource( );
			if ( source != null )
			{
				URL url = source.getLocation( );
				pathBase = url.getPath( );

				if ( pathBase.endsWith( "bin/" ) ) //$NON-NLS-1$
					pathBase = pathBase.substring( 0, pathBase.length( ) - 4 );
				if ( pathBase.endsWith( "bin" ) ) //$NON-NLS-1$
					pathBase = pathBase.substring( 0, pathBase.length( ) - 3 );
			}
		}

		pathBase = pathBase + getTestFolder( );
		String className = this.getClass( ).getName( );
		int lastDotIndex = className.lastIndexOf( "." ); //$NON-NLS-1$
		className = className.substring( 0, lastDotIndex );
		className = pathBase + className.replace( '.', '/' );

		return className;
	}

	/**
	 * Returns a list of class objects under the given package.
	 * 
	 * @param packageName
	 *            a given Java package
	 * @return a list of class objects under the given package.
	 */

	protected List getClasses( String packageName )
	{
		List classes = new ArrayList( );

		String pkgFolder = getClassFolder( )
				+ "/" + packageName.substring( packageName.lastIndexOf( '.' ) + 1 ); //$NON-NLS-1$
		File directory  = new File( pkgFolder );

		if ( directory.exists( ) )
		{
			// Get the list of the files contained in the package
			String[] files = directory.list( );
			for ( int i = 0; i < files.length; i++ )
			{
				// we are only interested in .class files
				if ( files[i].endsWith( ".java" ) ) //$NON-NLS-1$
				{
					// removes the .class extension

					classes.add( packageName + "."
							+ files[i].substring( 0, files[i].length( ) - 5 ) );
				}
			}
		}

		return classes;
	}
	
	/*
	 * A null-test for suppressing warnings in the build
	 */
	public void testNullTest( )
	{
		assertTrue( true );
	}
	
}
