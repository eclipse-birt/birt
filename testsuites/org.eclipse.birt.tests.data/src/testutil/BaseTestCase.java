/*******************************************************************************
 * Copyright (c) 2004,2005 Actuate Corporation.
 *
 * This program and the accompanying materials are made available under the
 * terms of the Eclipse Public License 2.0 which is available at
 * https://www.eclipse.org/legal/epl-2.0/.
 *
 * SPDX-License-Identifier: EPL-2.0
 *
 *
 * Contributors:
 *  Actuate Corporation  - initial API and implementation
 *******************************************************************************/

package testutil;

import java.io.BufferedReader;
import java.io.File;
import java.io.FileOutputStream;
import java.io.FileReader;
import java.io.IOException;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.io.PrintStream;
import java.net.URL;
import java.security.CodeSource;
import java.security.ProtectionDomain;

import org.mozilla.javascript.Context;
import org.mozilla.javascript.ImporterTopLevel;
import org.mozilla.javascript.Scriptable;

import junit.framework.TestCase;

/**
 * Common base class for all Dte test cases
 */
abstract public class BaseTestCase extends TestCase {

	protected PrintStream testOut;

	/** Top level Javascript scope */
	protected Context jsContext;
	protected Scriptable jsScope;

	// private static final String TEST_FOLDER = "test";
	private static final String OUTPUT_FOLDER = "output";
	private static final String INPUT_FOLDER = "input";
	private static final String GOLDEN_FOLDER = "golden";
	private static final String TEST_FOLDER = "src";

	private String classFolder;
	// private TimeZone defaultZone ;

	/*
	 * @see junit.framework.TestCase#setUp()
	 */
	@Override
	protected void setUp() throws Exception {
		super.setUp();

		// Create test output file
		// We must make sure this folder will be created successfully
		// before we do next job.
		openOutputFolder();
		openOutputFile();

		// Create top-level Javascript scope
		jsContext = Context.enter();
		jsScope = new ImporterTopLevel(jsContext);

		// Add JS functions testPrint and testPrintln for scripts to write
		// to output file
		jsScope.put("_testCase", jsScope, this);
		jsContext.evaluateString(jsScope, "function testPrint(str) { _testCase.testPrint(str); }; "
				+ "function testPrintln(str) { _testCase.testPrintln(str); }; ", "BaseTestCase.setUp", 1, null);
	}

	/*
	 * @see junit.framework.TestCase#tearDown()
	 */
	@Override
	protected void tearDown() throws Exception {
		Context.exit();
		closeOutputFile();
		super.tearDown();
	}

	/**
	 * Get input folder
	 *
	 * @return input folder
	 */
	protected File getInputFolder() {
		return new File(getBaseFolder(), INPUT_FOLDER);
	}

	/**
	 * Get base folder
	 *
	 * @return base folder
	 */
	private File getBaseFolder() {
		if (classFolder == null) {
			String pathBase = null;

			ProtectionDomain domain = this.getClass().getProtectionDomain();
			if (domain != null) {
				CodeSource source = domain.getCodeSource();
				if (source != null) {
					URL url = source.getLocation();
					pathBase = url.getPath();

					if (pathBase.endsWith("bin/")) { //$NON-NLS-1$
						pathBase = pathBase.substring(0, pathBase.length() - 4);
					}
					if (pathBase.endsWith("bin")) { //$NON-NLS-1$
						pathBase = pathBase.substring(0, pathBase.length() - 3);
					}
				}
			}

			pathBase = pathBase + TEST_FOLDER + "/";
			classFolder = pathBase.substring(1);
		}

		String className = this.getClass().getName();
		int lastDotIndex = className.lastIndexOf("."); //$NON-NLS-1$
		className = className.substring(0, lastDotIndex);
		className = classFolder + className.replace('.', '/');

		return new File(className);
	}

	/** return input folder */
	protected InputStream getInputFolder(String dataFileName) {
		InputStream result = this.getClass().getResourceAsStream(INPUT_FOLDER + "/" + dataFileName);
//		if ( result == null )
//		{
//			try
//			{
//
//				FileWriter writter = new FileWriter(
//						new File( "C:\\abcddd.txt" ) );
//				writter.write( String.valueOf( System.currentTimeMillis( ) ) );
//				writter.write( INPUT_FOLDER + File.separator + dataFileName );
//				writter.close( );
//				System.out.println( "Hello" );
//
//			}
//			catch ( FileNotFoundException e )
//			{
//				// TODO Auto-generated catch block
//				e.printStackTrace( );
//			}
//			// (this.getClass().getResource( INPUT_FOLDER + File.separator +
//			// dataFileName ).toString(
//			catch ( IOException e )
//			{
//				// TODO Auto-generated catch block
//				e.printStackTrace( );
//			}
//		}

		return result;

	}

	/** return output folder */
	private File getOutputFolder() {
		return new File(new File(System.getProperty("java.io.tmpdir")), OUTPUT_FOLDER);
	}

	/** return golder folder */
	/*
	 * private File getGoldenFolder( ) { return new File( getBaseFolder( ),
	 * GOLDEN_FOLDER ); }
	 */

	/** open output folder */
	private void openOutputFolder() {
		File outputFolder = getOutputFolder();
		if (!outputFolder.exists()) {
			outputFolder.mkdir();
		}
	}

	/** Opens defalt test output file. File name is ClassName.TestName.txt */
	protected void openOutputFile() throws IOException {
		File outputFile = new File(getOutputFolder(), getOutputFileName());
		testOut = new PrintStream(new FileOutputStream(outputFile, false));
	}

	/** close ouput file stream */
	protected void closeOutputFile() throws IOException {
		if (testOut != null) {
			testOut.close();
			testOut = null;
		}
	}

	/** return default output file name */
	private String getOutputFileName() {
//		FileWriter fw;
//		String className = "KK";
//		try
//		{
//			fw = new FileWriter( new File( "C:\\abc2.txt" ) );
//
//			fw.write( "ABC: " );
//
//			className = this.getClass( ).getName( );
//			fw.write( className + "\n" );
//			int lastDotIdx = className.lastIndexOf( '.' );
//			if ( lastDotIdx >= 0 )
//				className = className.substring( lastDotIdx + 1 );
//			fw.write( className + "\n" );
//			fw.write( this.getName( ) + "\n" );
//			fw.close( );
//		}
//		catch ( IOException e )
//		{
//			// TODO Auto-generated catch block
//			e.printStackTrace( );
//		}
//		return className + "." + this.getName( ) + ".txt";
		String className = this.getClass().getName();
		int lastDotIdx = className.lastIndexOf('.');
		if (lastDotIdx >= 0) {
			className = className.substring(lastDotIdx + 1);
		}
		return className + "." + this.getName() + ".txt";
	}

	/**
	 * Locates the folder where the unit test java source file is saved.
	 *
	 * @return the path where the test java source file locates.
	 */
	/*
	 * private File getBaseFolder( ) { if ( classFolder == null ) { String pathBase
	 * = null;
	 *
	 * ProtectionDomain domain = this.getClass( ).getProtectionDomain( ); if (
	 * domain != null ) { CodeSource source = domain.getCodeSource( ); if ( source
	 * != null ) { URL url = source.getLocation( ); pathBase = url.getPath( );
	 *
	 * if ( pathBase.endsWith( "bin/" ) ) //$NON-NLS-1$ pathBase =
	 * pathBase.substring( 0, pathBase.length( ) - 4 ); if ( pathBase.endsWith(
	 * "bin" ) ) //$NON-NLS-1$ pathBase = pathBase.substring( 0, pathBase.length( )
	 * - 3 ); } }
	 *
	 * pathBase = pathBase + TEST_FOLDER + "/"; classFolder = pathBase.substring( 1
	 * ); }
	 *
	 * String className = this.getClass( ).getName( ); int lastDotIndex =
	 * className.lastIndexOf( "." ); //$NON-NLS-1$ className = className.substring(
	 * 0, lastDotIndex ); className = classFolder + className.replace( '.', '/' );
	 *
	 * return new File( className ); }
	 *
	 *//**
		 * Asserts that output file matches the golden file. Default file name for
		 * current test case is used for both files
		 */
	protected void checkOutputFile() throws IOException {
		if (testOut != null) {
			testOut.flush();
		}

		String name = getOutputFileName();
		checkOutputFile(name, name);
	}

	/**
	 * Asserts that output file matches the golden file.
	 */
	private void checkOutputFile(String goldenFileName, String outputFileName) throws IOException {
		// FileWriter fw = new FileWriter( new File( "C:\\abc1.txt" ) );
		// fw.write( "ABC: " );
		// fw.write( goldenFileName );
		// fw.close( );
		InputStream golden = this.getClass().getResourceAsStream(GOLDEN_FOLDER + "/" + goldenFileName);
		File outputFile = new File(getOutputFolder(), outputFileName);
		assertTrue(compareTextFile(golden, outputFile));
	}

	/**
	 * compare two text file. The comparasion will ignore the line containing
	 * "modificationDate".
	 *
	 * @param goldenFileName the 1st file name to be compared.
	 * @param outputFileName the 2nd file name to be compared.
	 * @return True if two text file is same line by line
	 */
	private boolean compareTextFile(InputStream golden, File outputFile) throws IOException {
		boolean same = true;

		InputStreamReader readerA = new InputStreamReader(golden);
		FileReader readerB = new FileReader(outputFile);
		BufferedReader lineReaderA = new BufferedReader(readerA);
		BufferedReader lineReaderB = new BufferedReader(readerB);

		String strA = lineReaderA.readLine().trim();
		String strB = lineReaderB.readLine().trim();
		while (strA != null && strB != null) {
			same = strA.trim().equals(strB.trim());
			if (!same) {
				break;
			}

			strA = lineReaderA.readLine();
			strB = lineReaderB.readLine();
		}
		same = strA == null && strB == null;
		golden.close();
		readerA.close();
		readerB.close();
		lineReaderA.close();
		lineReaderB.close();

		return same;
	}

	/** print to console and stream */
	public void testPrint(String str) {
		System.out.print(str);
		if (testOut != null) {
			testOut.print(str);
		}
	}

	/** println to console and stream */
	public void testPrintln(String str) {
		System.out.println(str);
		if (testOut != null) {
			testOut.println(str);
		}
	}

}
