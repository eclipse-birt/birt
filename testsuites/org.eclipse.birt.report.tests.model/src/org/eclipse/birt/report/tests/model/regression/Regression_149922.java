/*******************************************************************************
 * Copyright (c) 2004 Actuate Corporation.
 *
 * This program and the accompanying materials are made available under the
 * terms of the Eclipse Public License 2.0 which is available at
 * https://www.eclipse.org/legal/epl-2.0/.
 *
 * SPDX-License-Identifier: EPL-2.0
 *
 ******************************************************************************/

package org.eclipse.birt.report.tests.model.regression;

import org.eclipse.birt.report.model.api.ElementFactory;
import org.eclipse.birt.report.model.api.ImageHandle;
import org.eclipse.birt.report.model.api.elements.DesignChoiceConstants;
import org.eclipse.birt.report.tests.model.BaseTestCase;

/**
 * <b>Regression description:</b>
 * <p>
 * Image builder - URI and file source type are treated as same
 * <p>
 * The image item is special item which can access both the global resources and
 * local resources. In the BIRT specification, we can find that the image can
 * access the resources through file or URI expression:
 * <p>
 * Accessing file images:
 *
 * <pre>
 *        &lt;image id=&quot;4&quot;&gt;
 *            &lt;property name=��source��&gt;file&lt;/property&gt;
 *            &lt;expression name=&quot;uri&quot;&gt;&quot;pict 103.jpg&quot;&lt;/expression&gt;
 *         &lt;/image&gt;
 * </pre>
 *
 * Accessing URI images
 *
 * <pre>
 *           &lt;image id=&quot;5&quot;&gt;
 *              &lt;property name=��source��&gt;url&lt;/property&gt;
 *               &lt;expression
 *   name=&quot;uri&quot;&gt;&quot;http://www.google.com/intl/en/images/logo.gif&quot;&lt;/expression&gt;
 *   &lt;/image&gt;
 * </pre>
 *
 * Unfortunately, the designer doesn��t distinguish those two kinds of
 * resources. The created report design is:
 *
 * <pre>
 *   &lt;image id=&quot;4&quot;&gt;
 *      &lt;expression name=&quot;uri&quot;&gt;&quot;pict 103.jpg&quot;&lt;/expression&gt;
 *   &lt;/image&gt;
 *
 *   &lt;image id=&quot;5&quot;&gt;
 *      &lt;expression
 *   name=&quot;uri&quot;&gt;&quot;http://www.google.com/intl/en/images/logo.gif&quot;&lt;/expression&gt;
 *   &lt;/image&gt;
 * </pre>
 *
 * <p>
 * That means both the local resources and the global resources are specified by
 * the URI expression.
 * <p>
 * <b>Test description:</b> Add two type of image FILE and URL to report, make
 * sure the types are correctly distinguished in the design file.
 * <p>
 * <p>
 */
public class Regression_149922 extends BaseTestCase {

	private final static String GOLDEN = "regression_149922.golden"; //$NON-NLS-1$
	private final static String OUTPUT = "regression_149922.out"; //$NON-NLS-1$

	@Override
	public void setUp() throws Exception {
		super.setUp();
		removeResource();
		copyGoldenToFile(GOLDEN_FOLDER + "/" + GOLDEN);

	}

	@Override
	public void tearDown() {
		removeResource();
	}

	/**
	 * @throws Exception
	 */

	public void test_regression_149922() throws Exception {
		this.createBlankDesign();
		ElementFactory factory = designHandle.getElementFactory();

		ImageHandle image = factory.newImage("img1"); //$NON-NLS-1$
		image.setTypeExpression(DesignChoiceConstants.IMAGE_REF_TYPE_FILE);
		image.setFile("\"images/pic1.jpg\""); //$NON-NLS-1$
		designHandle.getBody().add(image);

		ImageHandle image2 = factory.newImage("img1"); //$NON-NLS-1$
		image2.setTypeExpression(DesignChoiceConstants.IMAGE_REF_TYPE_URL);
		image2.setURL("\"http://www.google.com/intl/en/images/logo.gif\""); //$NON-NLS-1$
		designHandle.getBody().add(image2);

		String TempFile = this.genOutputFile(OUTPUT);
		designHandle.saveAs(TempFile);
		assertTrue(compareTextFile(GOLDEN, OUTPUT));

	}
}
