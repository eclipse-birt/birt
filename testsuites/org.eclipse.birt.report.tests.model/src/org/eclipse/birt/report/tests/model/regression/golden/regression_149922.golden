<?xml version="1.0" encoding="UTF-8"?>
<report xmlns="http://www.eclipse.org/birt/2005/design" version="3.2.17" id="1">
    <styles>
        <style name="crosstab" id="2">
            <property name="borderBottomColor">#CCCCCC</property>
            <property name="borderBottomStyle">solid</property>
            <property name="borderBottomWidth">1pt</property>
            <property name="borderLeftColor">#CCCCCC</property>
            <property name="borderLeftStyle">solid</property>
            <property name="borderLeftWidth">1pt</property>
            <property name="borderRightColor">#CCCCCC</property>
            <property name="borderRightStyle">solid</property>
            <property name="borderRightWidth">1pt</property>
            <property name="borderTopColor">#CCCCCC</property>
            <property name="borderTopStyle">solid</property>
            <property name="borderTopWidth">1pt</property>
        </style>
        <style name="crosstab-cell" id="3">
            <property name="borderBottomColor">#CCCCCC</property>
            <property name="borderBottomStyle">solid</property>
            <property name="borderBottomWidth">1pt</property>
            <property name="borderLeftColor">#CCCCCC</property>
            <property name="borderLeftStyle">solid</property>
            <property name="borderLeftWidth">1pt</property>
            <property name="borderRightColor">#CCCCCC</property>
            <property name="borderRightStyle">solid</property>
            <property name="borderRightWidth">1pt</property>
            <property name="borderTopColor">#CCCCCC</property>
            <property name="borderTopStyle">solid</property>
            <property name="borderTopWidth">1pt</property>
        </style>
    </styles>
    <body>
        <image name="img1" id="4">
            <property name="source">file</property>
            <expression name="uri">"images/pic1.jpg"</expression>
        </image>
        <image name="img11" id="5">
            <property name="source">url</property>
            <expression name="uri">"http://www.google.com/intl/en/images/logo.gif"</expression>
        </image>
    </body>
</report>
